source/posix/system_info.c

```diff
@@ -451,39 +456,65 @@ enum aws_platform_os aws_get_platform_build_os(void) {
 }
 #endif /* AWS_OS_APPLE */
 
-#if defined(__linux__) || defined(__unix__)
-static const uint64_t S_BYTES_TO_KILO_BYTES = 1024;
-#endif
-
-int aws_get_memory_usage(int64_t *output) {
+int aws_get_system_memory_usage(uint64_t *output) {
 // Get the Memory usage from Linux
 #if defined(__linux__) || defined(__unix__)
     struct sysinfo memory_info;
-    sysinfo(&memory_info);
+    int sysinfo_result = sysinfo(&memory_info);
+    if (sysinfo_result != 0) {
+        return aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
     uint64_t physical_memory_used = memory_info.totalram - memory_info.freeram;
     physical_memory_used *= memory_info.mem_unit;
-    // Return data in Kilobytes
-    physical_memory_used = physical_memory_used / S_BYTES_TO_KILO_BYTES;
     *output = physical_memory_used;
     return AWS_OP_SUCCESS;
-#endif
 
+#elif defined(__APPLE__)
+
+    vm_size_t page_size;
+    mach_port_t mach_port;
+    mach_msg_type_number_t count;
+    vm_statistics64_data_t vm_stats;
+
+    mach_port = mach_host_self();
+    count = sizeof(vm_stats) / sizeof(natural_t);
+    if (KERN_SUCCESS == host_page_size(mach_port, &page_size) &&
+        KERN_SUCCESS == host_statistics64(mach_port, HOST_VM_INFO, (host_info64_t)&vm_stats, &count)) {
+        *output =
+            (double)(((int64_t)vm_stats.active_count + (int64_t)vm_stats.inactive_count + (int64_t)vm_stats.wire_count) * (int64_t)page_size);
+        return AWS_OP_SUCCESS;
+    }
+    return aws_raise_error(AWS_ERROR_INVALID_STATE);
+
+#else
     // OS not supported? Just return an error and set the output to 0
     *output = 0;
-    aws_raise_error(AWS_ERROR_PLATFORM_NOT_SUPPORTED);
-    return AWS_OP_ERR;
+    return aws_raise_error(AWS_ERROR_PLATFORM_NOT_SUPPORTED);
+#endif
 }
 
-int aws_get_process_count(int64_t *output) {
+int aws_get_system_process_count(uint64_t *output) {
 // Get the process count from Linux
 #if defined(__linux__) || defined(__unix__)
     struct sysinfo system_info;
     sysinfo(&system_info);
-    *output = (int64_t)system_info.procs;
+    *output = (uint64_t)system_info.procs;
     return AWS_OP_SUCCESS;
-#endif
+
+#elif defined(__APPLE__)
+
+    pid_t pid_array[2048];
+    int pid_bytes = proc_listpids(PROC_ALL_PIDS, 0, pid_array, sizeof(pid_array));
+    if (pid_bytes == 0) {
```

Comment:
probably want to check for values less than 0 too

I can't find actual documentation for this function, but here's the [header](https://opensource.apple.com/source/xnu/xnu-2422.1.72/libsyscall/wrappers/libproc/libproc.h.auto.html) online and a nearby function has docs that say it returns -1 on error